oai_agents/agents/rl.py

```diff
@@ -301,21 +303,35 @@ def train_agents(self, total_train_timesteps, exp_name=None):
             if self.should_evaluate(steps=steps):
                 mean_training_rew = np.mean([ep_info["r"] for ep_info in self.learning_agent.agent.ep_info_buffer])                
                 if mean_training_rew >= self.best_training_rew:
+                    best_train_path, best_train_tag = self.save_agents(tag=CheckedPoints.BEST_TRAIN_REWARD)
+                    print(f'New best training score of {mean_training_rew} reached, model saved to {best_train_path}/{best_train_tag}')
                     self.best_training_rew = mean_training_rew
-                mean_reward, rew_per_layout = self.evaluate(self.learning_agent, timestep=self.learning_agent.num_timesteps)
+                if mean_training_rew <= self.worst_training_rew:
+                    worst_train_path, worst_train_tag = self.save_agents(tag=CheckedPoints.WORST_TRAIN_REWARD)
+                    print(f'Now worst training score of {mean_training_rew} reached, model saved to {worst_train_path}/{worst_train_tag}')
+                    self.worst_training_rew = mean_training_rew
 
+                mean_reward, rew_per_layout = self.evaluate(self.learning_agent, timestep=self.learning_agent.num_timesteps)
+                
                 if self.fcp_ck_rate:
                     if self.learning_agent.num_timesteps // self.fcp_ck_rate > (len(self.ck_list) - 1):
                         path, tag = self.save_agents(tag=f'ck_{len(self.ck_list)}_rew_{mean_reward}')
                         self.ck_list.append((rew_per_layout, path, tag))
 
                 if mean_reward >= self.best_score:
-                    best_path, best_tag = self.save_agents(tag='best')
-                    print(f'New best score of {mean_reward} reached, model saved to {best_path}/{best_tag}')
+                    best_path, best_tag = self.save_agents(tag=CheckedPoints.BEST_EVAL_REWARD)
+                    print(f'New best evaluation score of {mean_reward} reached, model saved to {best_path}/{best_tag}')
                     self.best_score = mean_reward
+                if mean_reward <= self.worst_score:
+                    worst_path, worst_tag = self.save_agents(tag=CheckedPoints.WORST_EVAL_REWARD)
```

Comment:
What is WORST_EVAL_REWARD used for

Reply:
For adversary, we need three checkpoints: worst_eval, best_train, and final_model.
adversary of worst_eval: brings the team performance down to the bottom.
adversary of best_train: brings its own training reward to the peak. 
adversary of final_model: the last checkpoints we have.

Reply:
From the experiments, the final_model is not usually either best_train or worst_eval, but usually close to best_train.

I have tested final_model and worst_eval with single-adversary-play. final_model works fine, but worst_eval may treat our agents too hard, and even make them to be passively aggressive. In a scenario, a well trained adversary tend to stop and not allow human-agents to pass if we trained adversary with worst_eval adversary.